common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +119,438 @@ public Collation(
+      protected final int collationId;
+
+      protected CollationSpec(
+          String locale,
+          CaseSensitivity caseSensitivity,
+          AccentSensitivity accentSensitivity,
+          CaseConversion caseConversion) {
+        this.locale = locale;
+        this.caseSensitivity = caseSensitivity;
+        this.accentSensitivity = accentSensitivity;
+        this.caseConversion = caseConversion;
+        this.collationId = getCollationId();
+      }
+
+      private static final Map<Integer, Collation> collationMap = new ConcurrentHashMap<>();
+
+      public static Collation fetchCollation(int collationId) throws SparkException {
+        if (collationMap.containsKey(collationId)) {
+          return collationMap.get(collationId);
+        } else {
+          CollationSpec spec;
+          int implementationProviderOrdinal =
+            (collationId >> IMPLEMENTATION_PROVIDER_OFFSET) & IMPLEMENTATION_PROVIDER_MASK;
+          if (implementationProviderOrdinal >= ImplementationProvider.values().length) {
+            throw SparkException.internalError("Invalid implementation provider");
+          } else {
+            ImplementationProvider implementationProvider = ImplementationProvider.values()[
+              implementationProviderOrdinal];
+            if (implementationProvider == ImplementationProvider.UTF8_BINARY) {
+              spec = CollationSpecUTF8Binary.fromCollationId(collationId);
+            } else if (implementationProvider == ImplementationProvider.ICU) {
+              spec = CollationSpecICU.fromCollationId(collationId);
+            } else {
+              throw SparkException.internalError("Cannot instantiate indeterminate collation");
+            }
+            Collation collation = spec.buildCollation();
+            collationMap.put(collationId, collation);
+            return collation;
+          }
+        }
+      }
+
+      public static int collationNameToId(String originalCollationName) throws SparkException {
+        String collationName = originalCollationName.toUpperCase();
+        try {
+          if (collationName.startsWith("UTF8_BINARY")) {
+            return CollationSpecUTF8Binary.collationNameToId(collationName);
+          } else {
+            return CollationSpecICU.collationNameToId(collationName);
+          }
+        } catch (SparkException e) {
+          throw new SparkException(
+            "COLLATION_INVALID_NAME",
+            SparkException.constructMessageParams(Map.of("collationName", originalCollationName)),
+            e);
+        }
+      }
+
+      protected static int parseSpecifiers(String specString) throws SparkException {
+        int specifiers = 0;
+        String[] parts = specString.split("_");
+        for (String part : parts) {
+          if (!part.isEmpty()) {
+            if (part.equals("UNSPECIFIED") || part.equals("INDETERMINATE")) {
+              throw new SparkException(part + " collation specifier reserved for internal use");
```

Comment:
Why is this case not combined with the default branch were we throw an error? Also one more question I have what happens if a user defines for example UNICODE_CI_CI_CI, do we allow this redundancy?

Reply:
Added tests to cover such problematic edge cases.